### PR TITLE
client API: deprecate opengl-cb API and introduce a replacement API

### PR DESCRIPTION
**Pull request by wm4**
The purpose of the new API is to make it useable with other APIs than
OpenGL, especially D3D11 and vulkan. In theory it's now possible to
support other vo_gpu backends, as well as backends that don't use the
vo_gpu code at all.

This also aims to get rid of the dumb mpv_get_sub_api() function. The
life cycle of the new mpv_render_context is a bit different from
mpv_opengl_cb_context, and you explicitly create/destroy the new
context, instead of calling init/uninit on an object returned by
mpv_get_sub_api().

In other to make the render API generic, it's annoyingly EGL style, and
requires you to pass in API-specific objects to generic functions. This
is to avoid explicit objects like the internal ra API has, because that
sounds more complicated and annoying for an API that's supposed to never
change.

The opengl_cb API will continue to exist for a bit longer, but
internally there are already a few tradeoffs, like reduced
thread-safety.

Mostly untested. Seems to work fine with mpc-qt.

CC @Akemi not sure if this breaks the swift backend

### CODE REVIEW

**Comment by wm4** [comment]
Also added an unrelated vapoursynth fix.

**Comment by Akemi** [comment]
yeah it [deadlocks](https://0x0.st/sT7P.txt), probably because i call API functions from main.c (cocoa_set_mpv_handle) in sync from the main thread. that was the reason i had to introduce a none-locking mp_get_sub_api function.

**Comment by Akemi** [comment]
just commenting the lock in `mp_set_main_render_context` makes it work again, but yeah that is not what you would want.

File: wscript_build.py [hunk 1]
```diff
@@ -161,11 +161,11 @@ def swift(task):
 
     if ctx.dependency_satisfied('macos-cocoa-cb'):
         swift_source = [
-            ( "video/out/cocoa_cb_common.swift" ),
-            ( "video/out/cocoa-cb/window.swift" ),
+            ( "osdep/macOS_mpv_helper.swift" )
```
Comment (Akemi) [inline wscript_build.py]: you missed a `,` (comma) here and there is an extra one at the end of the list.

**Comment by Akemi** [comment]
[ported](https://github.com/Akemi/mpv/commits/cocoa-cb_libmpv2) cocoa-cb to the new API.

File: libmpv/render.h [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+     * Type: int*
+     * This implies the depth of the surface passed to the render function in
+     * bits per channel. If omitted or set to 0, the renderer will assume 8.
+     * Typically used to control dithering.
+     */
+    MPV_RENDER_PARAM_DEPTH = 5,
+    /**
+     * ICC profile blob. Valid for mpv_render_context_set_parameter().
+     * Type: mpv_byte_array*
+     * Set an ICC profile for use with the "icc-profile-auto" option. (If this
```
Comment (Akemi) [inline libmpv/render.h]: an extra `this` or `the`.

File: video/out/vo.c [hunk 3]
```diff
@@ -185,7 +183,7 @@ static bool get_desc(struct m_obj_desc *dst, int index)
         .options = vo->options,
         .options_prefix = vo->options_prefix,
         .global_opts = vo->global_opts,
-        .hidden = vo->encode || !strcmp(vo->name, "opengl-cb"),
+        .hidden = vo->encode || !strcmp(vo->name, "libmpv"),
```
Comment (Akemi) [inline video/out/vo.c]: some people complained that `mpv --vo=help` won't list opengl-cb/libmpv with cocoa-cb enabled even though it is valid in that case. should we change this?

Comment (wm4) [reply video/out/vo.c]: OK, removed that.

**Comment by wm4** [comment]
Does anyone think mpv_render_param should be renamed to something else and be moved to client.h, just so it can be reused for something else in the future? Should the MPV_RENDER_PARAM_ enum receive the same treatment? Should the params maybe be strings instead of a magic number? These are pretty much questions about taste, but as creator of the API I can't really tell what is nicer (I've made my choices).

File: video/out/libmpv.h [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    const struct render_backend_fns *fns;
+
+    // Set on init, immutable afterwards.
+    int driver_caps;
+    struct mp_hwdec_devices *hwdec_devs;
+
+    void *priv;
+};
+
+// Generic backend for rendering via libmpv. This is corresponds to vo/vo_driver,
```
Comment (kevmitch) [inline video/out/libmpv.h]: extraneous "is"

File: video/out/libmpv.h [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+// except for rendering via the mpv_render_*() API. (As a consequence it's as
+// generic.) Like with VOs, one backend can support multiple underlying GPU APIs.
```
Comment (kevmitch) [inline video/out/libmpv.h]: as possible?

**Comment by kevmitch** [comment]
> Does anyone think mpv_render_param should be renamed to something else and be moved to client.h, just so it can be reused for something else in the future? Should the MPV_RENDER_PARAM_ enum receive the same treatment? Should the params maybe be strings instead of a magic number? These are pretty much questions about taste, but as creator of the API I can't really tell what is nicer (I've made my choices).

I'm not an API user, but my thoughts are that it would probably be better to generalize render params rather than worry about breaking the api in the future. I guess the enum would have to move with it. 
I like the enum because it makes typos detectable by the compiler.

**Comment by wm4** [comment]
Updated.

I now decided not to make the types/enums I mentioned generic, because they're still pretty trivial, and I think making them generic would probably be more awkward. Mostly I don't want to move the enum with all its members to client.h.

As mentioned, another possibility would be using strings instead of enums, but of course normally all string constants would have `#define`s for each them.

**Comment by kevmitch** [comment]
Would be nice to hear from some more API users, even if it's just "OK". @cmdrkotori, @gnome-mpv, @kanishka-linux, @lhc70000, @sonald, @varianus, @kokoko3k, @jeeb, @sfan5. @xyzz.

**Comment by kevmitch** [comment]
forgot @deltabeard

**Comment by deltabeard** [comment]
I'm happy with these changes.

**Comment by sfan5** [comment]
mpv-android no longer uses `opengl-cb`, these API changes look fine to me though.

**Comment by sonald** [comment]
looks good to me too.

**Comment by leavittx** [comment]
am I getting it right that although d3d11 render API integration is now possible in theory with this merged in, but it's not implemented yet? (ideally I would like to be able to get a native handle of a texture with video frames rendered to it by libmpv -- to be able to use it in the host application afterwards)

**Comment by wm4** [comment]
It's not implemented yet. d3d11 and Vulkan would be potential new backends. I think with d3d11 we have the problem that it's a very stateful API (similar to GL), and it's not clear how it can be avoided that application/libmpv break each other by setting unexpected state.

**Comment by leavittx** [comment]
I wonder would it be possible to use d3d11 backend without any opengl involved at all? I.e. no dx11/gl interop or ANGLE. Just pure dx11 code path. If yes, there is possibility to have d3d11 backend nicely wrapped into this API, I assume.

It's tricky to conclude that from the code itself, and I haven't managed to build it with debug symbols on windows yet (any hints on that btw?)

**Comment by wm4** [comment]
vo_gpu already has a pure d3d11 backend, and this new API is about making it accessible. The last part isn't implemented yet (needs glue code).

**Pull request merged**